test_runner.py

```diff
@@ -0,0 +1,132 @@
+from test_cli import TestModule
+
+import argparse
+import random
+import string
+import subprocess
+import sys
+
+
+class TestRunner(object):
+    _CODALAB_SERVICE_SCRIPT = 'codalab_service.py'
+
+    @staticmethod
+    def _docker_exec(command):
+        return 'docker exec -it codalab_rest-server_1 /bin/bash -c "{}"'.format(command)
+
+    @staticmethod
+    def _create_temp_instance(name):
+        print('Creating another CodaLab instance for testing...')
+
+        def get_free_ports(num_ports):
+            import socket
+
+            socks = [socket.socket(socket.AF_INET, socket.SOCK_STREAM) for _ in range(num_ports)]
+            for s in socks:
+                s.bind(("", 0))
+            ports = [str(s.getsockname()[1]) for s in socks]
+            for s in socks:
+                s.close()
+            return ports
+
+        rest_port, http_port, mysql_port = get_free_ports(3)
+        instance = 'http://rest-server:%s' % rest_port
+        try:
+            subprocess.check_call(
+                ' '.join(
+                    [
+                        'python3',
+                        TestRunner._CODALAB_SERVICE_SCRIPT,
+                        'start',
+                        '--instance-name %s' % name,
+                        '--rest-port %s' % rest_port,
+                        '--http-port %s' % http_port,
+                        '--mysql-port %s' % mysql_port,
+                        '--version %s' % version,
+                        '--services default',
+                    ]
+                ),
+                shell=True,
+            )
+        except subprocess.CalledProcessError as ex:
+            print('Temp instance exception: %s' % ex.output)
+            raise
+
+        return instance
+
+    def __init__(self, instance, tests):
+        self.instance = instance
+        self.temp_instance_name = 'temp-instance%s' % ''.join(
+            random.choice(string.digits) for _ in range(8)
+        )
+        self.temp_instance = TestRunner._create_temp_instance(self.temp_instance_name)
```

Comment:
We should only do this if `tests` require it...should be an annotation of the test.

Reply:
I added a check for this. I now only create a second instance if we are running `default`, `all` or if one of the test modules is `copy`.